app-android/src/androidTest/kotlin/care/data4life/integration/app/crud/BaseCrudSdkTest.kt

```diff
@@ -0,0 +1,224 @@
+        }
+
+        return loadedItems
+    }
+
+    abstract suspend fun callFetchByType(): List<Result<Record<T>>>
+    abstract fun assertFetchByType(expected: T, actual: Record<T>)
+
+    private suspend fun update(recordIds: List<String>, items: List<T>) {
+        // Given
+
+        // When
+        val results: List<Result<Record<T>>> = callUpdate(recordIds, items)
+
+        // Then
+        assertEquals(items.size, results.size)
+        assertResults(results, "Update") { index, record ->
+            assertUpdate(items[index], record)
+        }
+    }
+
+    abstract suspend fun callUpdate(recordIds: List<String>, items: List<T>): List<Result<Record<T>>>
+    abstract fun assertUpdate(expected: T, actual: Record<T>)
+
+    abstract fun supportsDownload(): Boolean
+    private suspend fun download(recordIds: List<String>, items: List<T>) {
+        // Given
+
+        // When
+        val results: List<Result<Record<T>>> = callDownload(recordIds)
+
+        // Then
+        assertEquals(recordIds.size, results.size)
+        assertResults(results, "Download") { index, record ->
+            assertDownload(items[index], record)
+        }
+    }
+
+    abstract suspend fun callDownload(recordIds: List<String>): List<Result<Record<T>>>
+    abstract fun assertDownload(expected: T, actual: Record<T>)
+
+    private suspend fun delete(recordIds: List<String>) {
+        // Given
+
+        // When
+        val results: List<Result<String>> = callDelete(recordIds)
+
+        // Then
+        assertEquals(recordIds.size, results.size)
+        results.forEachIndexed { index, result ->
+            when (result) {
+                is Success -> assertEquals(recordIds[index], result.data)
+                is Failure -> fail("Delete records (${recordIds[index]}) failed: ${result.exception.message}")
+            }
+        }
+    }
+
+    abstract suspend fun callDelete(recordIds: List<String>): List<Result<String>>
+
+    private suspend fun cleanAccount() {
+        val results: List<Result<Record<T>>> = callFetchByType()
+
+        val ids = results.filter { it is Success }
+            .map {
+                getId((it as Success).data)!!
+            }
+
+        if (ids.isNotEmpty()) {
+            deleteRecords(ids)
+        }
+    }
+
+    private suspend fun deleteRecords(recordIds: List<String>) {
+        awaitLegacyListener<LegacyDeleteResult> { listener ->
+            testSubject.deleteRecords(recordIds, listener)
+        }
+    }
+
+    private fun assertResults(
+        results: List<Result<Record<T>>>,
+        operationName: String,
+        check: (Int, Record<T>) -> Unit
+    ) {
+        for ((index, result) in results.withIndex()) {
+            when (result) {
+                is Success -> {
+                    val record = result.data
+                    check(index, record)
+                }
+                is Failure -> fail("$operationName record failed: ${result.exception.message}")
```

Comment:
If you restructure the Result you could spare the when. To get [inspiration](https://github.com/bitPogo/pixalb/blob/bd1c9a26d84eca3b9b76ada839438e3e6f7c0c99/coroutine-helper/src/commonMain/kotlin/io/bitpogo/util/coroutine/result/Result.kt#L9)

Reply:
See above